### PR TITLE
Revamp speculative decoding!

### PR DESCRIPTION
**Pull request by EricLBuehler**
Planning to entail:
- PagedAttention support
- Normal & Full cache support
- Cleaner implementation

@cdoko I'm planning on a full rewrite of speculative decoding. If you have the time to help with this, I would be happy to add you as a collaborator!

### CODE REVIEW

**Comment by EricLBuehler** [comment]
```
cargo run --release --features ... '--' -i toml -f speculative.toml
```

**Comment by EricLBuehler** [comment]
> Yes, I'd be happy to help!

Thanks! I sent you a collaborator invite.

**Comment by EricLBuehler** [comment]
@cdoko I sent you a collaborator invite, let me know if you have any trouble accessing the repo.

> From my reading, I thought the rejection sampling step would compare the probabilities of the target model and draft model for the token produced by the draft model. However, the code appears to sample all logits from the target model to generate 'gamma' number of tokens, and then compares these sampled tokens to the draft model's output. The acceptance criteria seems to be based on a match between the sampled tokens and the draft model's tokens.

I agree, the code certainly seems incorrect. I think it warrants a rewrite, what are your thoughts?

> Additionally, I'd like to propose adding device mapping support for speculative decoding. However, this would require implementing device mapping support for the regular cache first.

Yes, we should certainly implement that after it is working.

> I've also encountered a potential issue with the prefix cacher, as reported in https://github.com/EricLBuehler/mistral.rs/issues/1026. While testing speculative decoding, I've disabled the prefix cacher and modified the preop as mentioned in the issue.

Yes, PR #1035 disables the prefix cacher and reverts #1019 so it should *work* but no prefix caching until we find a permanent fix.
